genvalidity-hspec-optics/src/Test/Validity/Optics.hs

```diff
@@ -158,3 +174,115 @@ lensLaw3 l genB shrinkB genS shrinkS =
+--
+-- Example Usage:
+--
+-- prop> lensGettingProducesValidOnArbitrary ((_2) :: Lens (Double, Double) (Double, Double) Double Double)
+lensGettingProducesValidOnArbitrary ::
+       (Show s, Arbitrary s, Show b, Arbitrary b, Validity b)
+    => Lens s s b b
+    -> Property
+lensGettingProducesValidOnArbitrary l =
+    lensGettingProducesValidOnGen l arbitrary shrink
+
+-- | A property combinator to test whether getting values generated by given a generator via a lens on values generated by a given generator produces valid values.
+--
+-- > isValid (view l s)
+--
+-- Example Usage:
+--
+-- prop> lensGettingProducesValidOnGen ((_2) :: Lens (Double, Double) (Double, Double) Double Double) genValid shrinkValid
+lensGettingProducesValidOnGen ::
+       (Validity b, Show b, Show s)
+    => Lens s s b b
+    -> Gen s
+    -> (s -> [s])
+    -> Property
+lensGettingProducesValidOnGen l genS shrinkS =
+    forAllShrink genS shrinkS $ \s -> shouldBeValid $ view l s
+
+-- | A property combinator to test whether setting valid values via a lens on valid values produces valid values.
+--
+-- Example usage:
+--
+-- prop> lensSettingProducesValidOnValid ((_2) :: Lens (Double, Double) (Double, Double) Double Double)
+lensSettingProducesValidOnValid ::
+       (Show s, GenValid s, Show b, GenValid b, Show t, Validity t)
+    => Lens s t a b
+    -> Property
+lensSettingProducesValidOnValid l =
+    lensSettingProducesValidOnGen l genValid shrinkValid genValid shrinkValid
+
+-- | A property combinator to test whether setting unchecked values via a lens on unchecked values produces valid values.
+--
+-- Example usage:
+--
+-- prop> lensSettingProducesValid ((_2) :: Lens (Int, Int) (Int, Int) Int Int)
+lensSettingProducesValid ::
+       (Show s, GenUnchecked s, Show b, GenUnchecked b, Show t, Validity t)
+    => Lens s t a b
+    -> Property
+lensSettingProducesValid l =
+    lensSettingProducesValidOnGen
+        l
+        genUnchecked
+        shrinkUnchecked
+        genUnchecked
+        shrinkUnchecked
+
+-- | A property combinator to test whether setting arbitrary values via a lens on arbitrary values produces valid values.
+--
+-- Example usage:
+--
+-- prop> lensSettingProducesValidOnArbitrary ((_2) :: Lens (Double, Double) (Double, Double) Double Double)
+lensSettingProducesValidOnArbitrary ::
+       (Show s, Arbitrary s, Show b, Arbitrary b, Show t, Validity t)
+    => Lens s t a b
+    -> Property
+lensSettingProducesValidOnArbitrary l =
+    lensSettingProducesValidOnGen l arbitrary shrink arbitrary shrink
+
+-- | A property combinator to test whether setting values generated by given a generator via a lens on values generated by a given generator produces valid values.
+--
+-- > isValid (set l b s)
+--
+-- Example usage:
+--
+-- prop> lensSettingProducesValidOnGen ((_2) :: Lens (Double, Double) (Double, Double) Double Double) genValid shrinkValid genValid shrinkValid
+lensSettingProducesValidOnGen ::
```

Comment:
It sounds like this test should be included in lensSpec and lensSpecOnValid with the correct generators and such.

Reply:
Great idea!